Update audit log docs (#4919)
hammerandchisel/discord-api-docs: docs/resources/Audit_Log.md

## Audit Logs

> When an administrative action is performed in a guild, an entry is added to its audit log. Viewing audit logs requires the `VIEW_AUDIT_LOG` permission and can be fetched by apps using the [`GET /guilds/{guild.id}/audit-logs` endpoint](#DOCS_RESOURCES_AUDIT_LOG/get-guild-audit-log), or seen by users in the guild's **Server Settings**. All audit log entries are stored for 45 days.

Reviewer: aren't they stored for 90 days? i've always heard it was 90 and not 45

Author: hmmm when I was looking in the code it looked like it was 45, but let me check again (it's also going to change at some point in the future)

Reviewer 2: I believe it used to be 90 but was changed to 45 years ago

Author: It looks like only 45 at the moment—checked in some internal docs and verified in the client for a few different guilds.

| changes?    | array of [audit log change](#DOCS_RESOURCES_AUDIT_LOG/audit-log-change-object) objects                  | Changes made to the target_id                         |
| user_id     | ?snowflake                                                                                              | User or app that made the changes                     |
| id          | snowflake                                                                                               | ID of the entry                                       |
| action_type | [audit log event](#DOCS_RESOURCES_AUDIT_LOG/audit-log-entry-object-audit-log-events)                    | Type of action that occurred                          |
| options?    | [optional audit entry info](#DOCS_RESOURCES_AUDIT_LOG/audit-log-entry-object-optional-audit-entry-info) | Additional info for certain action types              |
| reason?     | string                                                                                                  | Reason for the change (1-512 characters)              |

> warn

> > For `APPLICATION_COMMAND_PERMISSION_UPDATE` events, the `target_id` is the app ID since the `changes` array may contain several permission changes for different individual entities

Reviewer 3: ```suggestion
> For `APPLICATION_COMMAND_PERMISSION_UPDATE` events, the `target_id` is the app ID since the `changes` array may contain several permission changes for different individual entities.
```

> > For `APPLICATION_COMMAND_PERMISSION_UPDATE` events, the `target_id` is the app ID since the `changes` array may contain several permission changes for different individual entities.

Reviewer 4: ```suggestion
> For `APPLICATION_COMMAND_PERMISSION_UPDATE` events, the `target_id` is the app or command ID since the `changes` array represents the entire `permissions` property on the [guild permissions](#DOCS_INTERACTIONS_APPLICATION_COMMANDS/application-command-permissions-object-guild-application-command-permissions-structure) object.
```

target_id is the same id used in the permissions update endpoint, which is usually a command id, but can also be application id for the "global" permissions entity. You also can not set more than one commands permissions anymore.

| GUILD_SCHEDULED_EVENT_DELETE          | 102   | Event was cancelled                                       | [Guild Scheduled Event](#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-object)                                                    |
| THREAD_CREATE                         | 110   | Thread was created in a channel                           | [Thread](#DOCS_RESOURCES_CHANNEL/thread-metadata-object)                                                                                       |
| THREAD_UPDATE                         | 111   | Thread was updated                                        | [Thread](#DOCS_RESOURCES_CHANNEL/thread-metadata-object)                                                                                       |
| THREAD_DELETE                         | 112   | Thread was deleted                                        | [Thread](#DOCS_RESOURCES_CHANNEL/thread-metadata-object)                                                                                       |

> | APPLICATION_COMMAND_PERMISSION_UPDATE | 121   | Permissions were updated for a command                    | [Command Permission](#DOCS_INTERACTIONS_APPLICATION_COMMANDS/application-command-permissions-object-application-command-permissions-structure) |

Reviewer 4: ```suggestion
| APPLICATION_COMMAND_PERMISSION_UPDATE | 121   | Permissions were updated for a command                    | [Command Permission](#DOCS_INTERACTIONS_APPLICATION_COMMANDS/application-command-permissions-object-guild-application-command-permissions-structure) |
```

The current link is for a single permission update, which while intuitively is the entity being updated, isn't the one actually represented by this entry.

| role_name          | string    | Name of the role if type is `"0"` (not present if type is `"1"`) | CHANNEL_OVERWRITE_CREATE & CHANNEL_OVERWRITE_UPDATE & CHANNEL_OVERWRITE_DELETE                                                     |
| type               | string    | Type of overwritten entity - role (`"0"`) or member (`"1"`)      | CHANNEL_OVERWRITE_CREATE & CHANNEL_OVERWRITE_UPDATE & CHANNEL_OVERWRITE_DELETE                                                     |

### Audit Log Change Object

Many audit log events include a `changes` array in their [entry object](#DOCS_RESOURCES_AUDIT_LOG/audit-log-entry-object-audit-log-entry-structure). The structure for the individual changes varies based on the event type and its changed objects, so apps shouldn't depend on a single pattern of handling audit log events.

###### Audit Log Change Structure

> warn
> `APPLICATION_COMMAND_PERMISSIONS_UPDATE` and `MEMBER_ROLE_UPDATE` events don't follow the same pattern as other audit log events. Details can be found in the [change key](#DOCS_RESOURCES_AUDIT_LOG/audit-log-change-object-audit-log-change-key) section.

> info

> > If `new_value` is not present in the change object while `old_value` is, it indicates that the changed property has been reset or set to `null`.

Reviewer 5: ```If `new_value` is not present in the change object while `old_value` is, it indicates that the changed property has been reset or set to `null`.```

Since this line now acts as the only documentation for the fact that any fields marked as nullable on their relevant object would never be `null` here, would it be worth also documenting that `old_value` won't be present if the field was originally `null`?

| Field      | Type                                                                            | Description                                                                                            |
| ---------- | ------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------ |
| new_value? | [mixed](#DOCS_RESOURCES_AUDIT_LOG/audit-log-change-object-audit-log-change-key) | New value of the key                                                                                   |
| old_value? | [mixed](#DOCS_RESOURCES_AUDIT_LOG/audit-log-change-object-audit-log-change-key) | Old value of the key                                                                                   |

> | key        | string                                                                          | Name of the changed entity (one of the [change keys](#DOCS_RESOURCES_AUDIT_LOG/audit-log-change-object-audit-log-change-key)) |

Reviewer 4: Is it worth noting the exception here somehow?

| [Channel Overwrite](#DOCS_RESOURCES_CHANNEL/overwrite-object)                                                                                    | `id`, `type`, `allow`, `deny`                                                                                                                                                                                                                                                                                                                                                                                                                                                          |

> | [Command Permission](#DOCS_INTERACTIONS_APPLICATION_COMMANDS/application-command-permissions-object-application-command-permissions-structure)\* | snowflake                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |

Reviewer 5: This seems a bit inconsistent, shouldn't the values for `old_value` and `new_value` objects be documented here for Command Permissions (rather than as an asterisks under the table), not the special casing for the value of the `key` field which would be a better fit documented on the Audit Log Change Structure object?
